src/enclave/enclave_init.c

```diff
@@ -8,205 +8,202 @@
 extern struct mpmcq __scheduler_queue;
 
-_Noreturn void __dls3(sgxlkl_app_config_t* conf, void* tos);
+_Noreturn void __dls3(elf64_stack_t* conf, void* tos);
 extern void init_sysconf(long nproc_conf, long nproc_onln);
 
-/* Copy the application setting/conf to enclave memory */
-static void __sgxlkl_enclave_copy_app_config(
-    sgxlkl_app_config_t* app_config,
-    const sgxlkl_config_t* sgxlkl_config)
+static void find_and_mount_disks()
 {
-    int i = 0, envc = 0;
-    char** envp = NULL;
-    size_t total_size = 0;
+    const sgxlkl_enclave_config_t* cfg = sgxlkl_enclave_state.config;
+    size_t n = cfg->num_mounts + 1;
 
-    app_config->argc = sgxlkl_config->argc;
+    sgxlkl_enclave_state_t* estate = &sgxlkl_enclave_state;
+    const sgxlkl_shared_memory_t* shm = &estate->shared_memory;
 
-    for (i = 0; i < app_config->argc; i++)
-        total_size += oe_strlen(sgxlkl_config->argv[i]) + 1;
+    estate->disk_state = oe_calloc(n, sizeof(sgxlkl_enclave_disk_state_t));
+    estate->num_disk_state = n;
 
-    envp = sgxlkl_config->argv + sgxlkl_config->argc + 1;
-    while (envp[envc] != NULL)
+    // root disk index
+    estate->disk_state[0].host_disk_index = 0;
+
+    for (int i = 0; i < cfg->num_mounts; i++)
     {
-        total_size += oe_strlen(envp[envc]) + 1;
-        envc++;
-    }
+        const sgxlkl_enclave_mount_config_t* cfg_disk = &cfg->mounts[i];
 
-    char* buf = oe_malloc((total_size + 1) * sizeof(char*));
-    app_config->argv =
-        oe_malloc((sgxlkl_config->argc + envc + 2) * sizeof(char*));
-    size_t remaining = total_size + 1;
+        if (strcmp(cfg_disk->destination, "/") == 0)
+            sgxlkl_fail("Error: root disk should not be in 'mounts'.\n");
 
-    char* p = buf;
-    for (i = 0; i < app_config->argc; i++)
-    {
-        app_config->argv[i] = p;
-        p += oe_snprintf(p, remaining, "%s", sgxlkl_config->argv[i]) + 1;
-        remaining -= p - app_config->argv[i];
+        bool found = false;
+        for (int j = 0; j < shm->num_virtio_blk_dev && !found; j++)
+        {
+            if (strcmp(cfg_disk->destination, shm->virtio_blk_dev_names[j]) ==
+                0)
+            {
+                estate->disk_state[i + 1].host_disk_index = j;
+                found = true;
+            }
+        }
+        if (!found)
+            sgxlkl_fail(
+                "Disk image for mount point '%s' has not been provided by "
+                "host.\n",
+                cfg_disk->destination);
     }
-    app_config->argv[i] = NULL;
 
-    app_config->envp = &app_config->argv[i + 1];
-    for (i = 0; i < envc; i++)
-    {
-        app_config->envp[i] = p;
-        p += oe_snprintf(p, remaining, "%s", envp[i]) + 1;
-        remaining -= p - app_config->envp[i];
-    }
-    app_config->envp[i] = NULL;
+    lkl_mount_disks(&cfg->root, cfg->mounts, cfg->num_mounts, cfg->cwd);
+}
 
-    app_config->cwd = oe_strdup(sgxlkl_config->cwd);
+// In internal OE header openenclave/internal/sgx/eeid_plugin.h
```

Comment:
There doesn't seem to be a clear policy yet (see also https://github.com/openenclave/openenclave/pull/3131). It was rather my mistake for not installing those headers as part of the original OE PR. I've since added more and I think only the eeid_verifier is missing, but will be there shortly.

There is a more general problem with OE internal headers for oe-libc and oe-corelibc though.

Reply:
OK, maybe add a TODO note to remove these declarations once that's fixed in OE.

Reply:
This whole section is gone, there'll be a re-layered equivalent soon.